api/core/key_test.go

```diff
@@ -0,0 +1,415 @@
+		name string
+		v    []byte
+		want Value
+	}{
+		{
+			name: `value: []byte{'f','o','o'}`,
+			v:    []byte{'f', 'o', 'o'},
+			want: Value{
+				Type:  BYTES,
+				Bytes: []byte{'f', 'o', 'o'},
+			},
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			//proto: func (k Key) Bytes(v []byte) KeyValue {
+			have := Key{}.Bytes(testcase.v)
+			if diff := cmp.Diff(testcase.want, have.Value); diff != "" {
+				t.Fatal(diff)
+			}
+		})
+	}
+}
+
+func TestInt(t *testing.T) {
+	WTYPE := INT64
+	if unsafe.Sizeof(int(42)) == 4 {
+		// switch the desired value-type depending on system int byte-size
+		WTYPE = INT32
+	}
+
+	for _, testcase := range []struct {
+		name string
+		v    int
+		want Value
+	}{
+		{
+			name: `value: int(42)`,
+			v:    int(42),
+			want: Value{
+				Type:  WTYPE,
+				Int64: int64(42),
+			},
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			//proto: func (k Key) Int(v int) KeyValue {
+			have := Key{}.Int(testcase.v)
+			if diff := cmp.Diff(testcase.want, have.Value); diff != "" {
+				t.Fatal(diff)
+			}
+		})
+	}
+}
+
+func TestUint(t *testing.T) {
+	WTYPE := UINT64
+	if unsafe.Sizeof(uint(42)) == 4 {
+		// switch the desired value-type depending on system int byte-size
+		WTYPE = UINT32
+	}
+
+	for _, testcase := range []struct {
+		name string
+		v    uint
+		want Value
+	}{
+		{
+			name: `value: uint(42)`,
+			v:    uint(42),
+			want: Value{
+				Type:   WTYPE,
+				Uint64: 42,
+			},
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			//proto: func (k Key) Uint(v uint) KeyValue {
+			have := Key{}.Uint(testcase.v)
+			if diff := cmp.Diff(testcase.want, have.Value); diff != "" {
+				t.Fatal(diff)
+			}
+		})
+	}
+}
+
+func TestDefined(t *testing.T) {
+	for _, testcase := range []struct {
+		name string
+		k    Key
+		want bool
+	}{
+		{
+			name: `Key Defined`,
+			k: Key{
+				registry.Variable{
+					Name: "foo",
+				},
+			},
+			want: true,
+		},
+		{
+			name: `Key not Defined`,
+			k:    Key{registry.Variable{}},
+			want: false,
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			//func (k Key) Defined() bool {
+			have := testcase.k.Defined()
+			if have != testcase.want {
+				t.Errorf("Want: %v, but have: %v", testcase.want, have)
+			}
+		})
+	}
+}
+
+func TestEmit(t *testing.T) {
+	for _, testcase := range []struct {
+		name string
+		v    Value
+		want string
+	}{
+		{
+			name: `bool`,
+			v: Value{
+				Type: BOOL,
+				Bool: true,
+			},
+			want: "true",
+		},
+		{
+			name: `int32`,
+			v: Value{
+				Type:  INT32,
+				Int64: 42,
+			},
+			want: "42",
+		},
+		{
+			name: `int64`,
+			v: Value{
+				Type:  INT64,
+				Int64: 42,
+			},
+			want: "42",
+		},
+		{
+			name: `uint32`,
+			v: Value{
+				Type:   UINT32,
+				Uint64: 42,
+			},
+			want: "42",
+		},
+		{
+			name: `uint64`,
+			v: Value{
+				Type:   UINT64,
+				Uint64: 42,
+			},
+			want: "42",
+		},
+		{
+			name: `float32`,
+			v: Value{
+				Type:    FLOAT32,
+				Float64: 42,
```

Comment:
nit: can you please change 42 to 42.1? same with FLOAT64.

Reply:
I did this initially, but after the type conversions the value came back as not exactly 42.1 but something like 42.100006542 😞 

42.0 works if that's acceptable

Reply:
I suppose I could loosen the test to only consider one sig-fig 🤷‍♂

Reply:
I tried with 42.1 and I got 42.1
```
=== RUN   TestEmit/float32
have: 42.1
=== RUN   TestEmit/float64
have: 42.1
```

There are many test case in Opencensus that uses XY.Z form for double. Is it specific to your environment/setup. You may want to make the change and submit it and then see if the test passes/fails on circle-ci.

Reply:
looks like it fails circle too 😞 

``` --- FAIL: TestFloat32 (0.00s)
    --- FAIL: TestFloat32/value:_float32(42.1) (0.00s)
        key_test.go:180:   core.Value{
              	... // 2 identical fields
              	Int64:   0,
              	Uint64:  0x00,
            - 	Float64: 42.1,
            + 	Float64: 42.099998474121094,
              	String:  "",
              	Bytes:   nil,
              }```

Reply:
So it's only a single float test that's failing it, so I'm setting everything to 42.1 except TestFloat32(), which I'm setting to 42.0

Reply:
Earlier I had only changed it for TestEmit. It works there but it doesn't work for TestFloat32. So lets keep your original commit.